### PR TITLE
=str #17200 Stop shard region when MemberRemoved

### PR DESCRIPTION
**Pull request by patriknw**

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #17200

**Comment by patriknw** [comment]
Updated with new solution. Two issues:

1) ShardRegion actor must stop itself when the node is shutting down,
   ie. when receiving MemberRemoved(selfAddress)

2) ShardCoordinator must not persist anything when the node is shutting
   down. MemberRemoved of other shard regions will trigger Terminated,
   which must not be persisted, because then the next coordinator will
   replay those events and end up in wrong state. This is a problem
   announced itself when using leaving as illustrated in the new test.

To solve the second issue I have added a new ClusterShuttingDown event
that is published before the MemberRemoved events. Note that Terminated
is triggered by MemberRemoved.

File: akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala [hunk 1]
```diff
@@ -170,6 +170,17 @@ object ClusterEvent {
   }
 
   /**
+   * This event is published when the cluster node is shutting down,
+   * before the final [[MemberRemoved]] events are published.
+   */
+  case object ClusterShuttingDown extends ClusterDomainEvent
```
Comment (rkuhn) [inline akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala]: we regard these as local-only messages, right?

File: akka-contrib/src/main/scala/akka/contrib/pattern/ClusterSharding.scala [hunk 2]
```diff
@@ -1331,6 +1337,18 @@ class ShardCoordinator(handOffTimeout: FiniteDuration, rebalanceInterval: Finite
       persistentState.regionProxies.foreach(context.watch)
       persistentState.regions.foreach { case (a, _) ⇒ context.watch(a) }
 
+    case ClusterShuttingDown ⇒
+      log.debug("Shutting down ShardCoordinator")
+      // can't stop because supervisor will start it again,
+      // it will soon be stopped when singleton is stopped
+      context.become(shuttingDown)
```
Comment (rkuhn) [inline akka-contrib/src/main/scala/akka/contrib/pattern/ClusterSharding.scala]: perhaps we should introduce `Actor.ignoringBehavior`?

File: akka-contrib/src/multi-jvm/scala/akka/contrib/pattern/ClusterShardingLeavingSpec.scala [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+      case GetLocations ⇒ sender() ! locations
+      case l: Locations ⇒ locations = l
+    }
+  }
+
+  val idExtractor: ShardRegion.IdExtractor = {
+    case m @ Ping(id) ⇒ (id, m)
+  }
+
+  val shardResolver: ShardRegion.ShardResolver = msg ⇒ msg match {
```
Comment (rkuhn) [inline akka-contrib/src/multi-jvm/scala/akka/contrib/pattern/ClusterShardingLeavingSpec.scala]: wouldn’t it work without the `msg => msg match ` part?

File: akka-contrib/src/multi-jvm/scala/akka/contrib/pattern/ClusterShardingLeavingSpec.scala [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+          region ! Ping(id)
+          id -> expectMsgType[ActorRef]
+        }).toMap
+        shardLocations ! Locations(locations)
+      }
+      enterBarrier("after-3")
+    }
+
+    "recover after leaving coordinator node" in within(30.seconds) {
+      runOn(second, third, fourth) {
```
Comment (rkuhn) [inline akka-contrib/src/multi-jvm/scala/akka/contrib/pattern/ClusterShardingLeavingSpec.scala]: this looks extraneous

**Comment by rkuhn** [comment]
LGTM

File: akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala (hunk 1 above)
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala]: yes, that is the intended usage. we don't have custom serialization or serialversionuid for any of these ClusterDomainEvent

File: akka-contrib/src/main/scala/akka/contrib/pattern/ClusterSharding.scala (hunk 2 above)
Comment (patriknw) [inline akka-contrib/src/main/scala/akka/contrib/pattern/ClusterSharding.scala]: nice to have category, added ticket #17257

File: akka-contrib/src/multi-jvm/scala/akka/contrib/pattern/ClusterShardingLeavingSpec.scala (hunk 3 above)
Comment (patriknw) [inline akka-contrib/src/multi-jvm/scala/akka/contrib/pattern/ClusterShardingLeavingSpec.scala]: indeed, copy pasta, replaced 3 occurrences

File: akka-contrib/src/multi-jvm/scala/akka/contrib/pattern/ClusterShardingLeavingSpec.scala (hunk 4 above)
Comment (patriknw) [inline akka-contrib/src/multi-jvm/scala/akka/contrib/pattern/ClusterShardingLeavingSpec.scala]: yeah :)

**Comment by patriknw** [comment]
final touch incorporates @rkuhn's comments

**Comment by drewhk** [comment]
LGTM

**Comment by rkuhn** [comment]
fixes #17200

**Comment by 2m** [comment]
LGTM

File: akka-contrib/src/multi-jvm/scala/akka/contrib/pattern/ClusterShardingLeavingSpec.scala [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+    "setup shared journal" in {
+      // start the Persistence extension
+      Persistence(system)
+      runOn(first) {
+        system.actorOf(Props[SharedLeveldbStore], "store")
+      }
+      enterBarrier("peristence-started")
+
+      runOn(first, second) {
```
Comment (patriknw) [inline akka-contrib/src/multi-jvm/scala/akka/contrib/pattern/ClusterShardingLeavingSpec.scala]: I had big trouble porting this PR to master. This test failed, and it turned out that this is wrong. It should initialize the SharedLeveldbJournal.setStore on all nodes. The reason it fails in master is the new rememberEntries feature, which requires persistence on all nodes, even if the feature is not used. I'll create a ticket about that.

**Pull request merged**